File: harness/cpp/harness.cpp
// Trace harness: captures median nanosecond costs for the six crime
// operations and four rescue operations used in the honestcode.software
// demo engine.
//
// Usage:
//   g++ -O2 -std=c++17 harness.cpp -o harness && ./harness

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <memory>
#include <numeric>
#include <string>
#include <unordered_map>
#include <stdexcept>
#include <vector>

constexpr int WARMUP = 200;
constexpr int RUNS = 1000;

using Clock = std::chrono::high_resolution_clock;

inline long long ns() {
    return std::chrono::duration_cast<std::chrono::nanoseconds>(
        Clock::now().time_since_epoch()).count();
}

// ═══════════════════════════════════════════════
// CRIME SCENE: mutable class with singletons
// ═══════════════════════════════════════════════

class CouponRegistry {
    static std::shared_ptr<CouponRegistry> instance;
public:
    static std::shared_ptr<CouponRegistry> getInstance() {
        if (!instance) instance = std::make_shared<CouponRegistry>();
        return instance;
    }
    static void reset() { instance.reset(); }
    double lookup(const std::string& code) const {
        return code == "SAVE10" ? 0.10 : 0;
    }
};
std::shared_ptr<CouponRegistry> CouponRegistry::instance;

class TaxService {
    static std::shared_ptr<TaxService> instance;
public:
    static std::shared_ptr<TaxService> getInstance() {
        if (!instance) instance = std::make_shared<TaxService>();
        return instance;
    }
    static void reset() { instance.reset(); }
    double calculate(const std::string& region, double taxable) const {
        return taxable * 0.08;
    }
};
std::shared_ptr<TaxService> TaxService::instance;

struct Item {
    std::string name;
    double price;
};

struct Order {
    std::vector<Item> items;
    double total = 0;
    double discount = 0;
    double tax = 0;
    std::string couponCode;
    long long updatedAt = 0;
};

struct CrimeResult {
    long long call, field, calc, single_, cache, time_;
};

CrimeResult crimeRun() {
    Order order;
    std::vector<Item> items = {{"Widget", 29.99}, {"Gadget", 39.99}, {"Doohickey", 19.99}};

    auto t0 = ns();
    order.items.insert(order.items.end(), items.begin(), items.end());
    auto callNs = ns() - t0;

    t0 = ns();
    order.couponCode = "SAVE10";
    order.discount = 0;
    auto fieldNs = ns() - t0;

    t0 = ns();
    order.total = std::accumulate(order.items.begin(), order.items.end(), 0.0,
        [](double sum, const Item& i) { return sum + i.price; });
    auto calcNs = ns() - t0;

    CouponRegistry::reset();
    TaxService::reset();
    t0 = ns();
    auto registry = CouponRegistry::getInstance();
    auto taxService = TaxService::getInstance();
    auto singleNs = (ns() - t0) / 2;

    t0 = ns();
    auto rate = registry->lookup(order.couponCode);
    order.discount = order.total * rate;
    auto cacheNs = ns() - t0;

    t0 = ns();
    order.updatedAt = ns();
    auto timeNs = ns() - t0;

    order.tax = taxService->calculate("NY", order.total - order.discount);

    return {callNs, fieldNs, calcNs, singleNs, cacheNs, timeNs};
}

// ═══════════════════════════════════════════════
// RESCUE: pure functions, flat data
// ═══════════════════════════════════════════════

struct OrderResult {
    double total, tax, subtotal;
};

struct FinalResult {
    double total, tax, subtotal, discount, grandTotal;
    std::string couponCode;
};

OrderResult calculateOrder(const std::vector<Item>& items, const std::string& region,
                           const std::unordered_map<std::string, double>& taxRates) {
    double total = std::accumulate(items.begin(), items.end(), 0.0,
        [](double sum, const Item& i) { return sum + i.price; });
    auto it = taxRates.find(region);
    double tax = total * (it != taxRates.end() ? it->second : 0);
    return {total, tax, total + tax};
}

FinalResult applyCoupon(const OrderResult& order, const std::string& code,
                        const std::unordered_map<std::string, double>& coupons) {
    auto it = coupons.find(code);
    double rate = it != coupons.end() ? it->second : 0;
    double discount = order.total * rate;
    return {order.total, order.tax, order.subtotal, discount, order.subtotal - discount, code};
}

struct RescueResult {
    long long call, arg, calc, ret;
};

RescueResult rescueRun() {
    std::vector<Item> items = {{"Widget", 29.99}, {"Gadget", 39.99}, {"Doohickey", 19.99}};
    std::unordered_map<std::string, double> taxRates = {{"NY", 0.08}, {"CA", 0.0725}};
    std::unordered_map<std::string, double> coupons = {{"SAVE10", 0.10}};

    auto t0 = ns();
    auto t1 = ns();
    auto callNs = t1 - t0;

    t0 = ns();
    std::string region = "NY";
    auto argNs = ns() - t0;

    t0 = ns();
    auto result = calculateOrder(items, region, taxRates);
    auto calcNs = ns() - t0;

    t0 = ns();
    auto final_ = applyCoupon(result, "SAVE10", coupons);
    auto retNs = ns() - t0;

    if (final_.grandTotal < 0) throw std::runtime_error("impossible");

    return {callNs, argNs, calcNs, retNs};
}

// ═══════════════════════════════════════════════
// HARNESS
// ═══════════════════════════════════════════════

long long median(std::vector<long long>& values) {
    std::sort(values.begin(), values.end());
    return values[values.size() / 2];
}

int main() {
    for (int i = 0; i < WARMUP; i++) { crimeRun(); rescueRun(); }

    std::vector<CrimeResult> crimeResults(RUNS);
    std::vector<RescueResult> rescueResults(RUNS);
    for (int i = 0; i < RUNS; i++) {
        crimeResults[i] = crimeRun();
        rescueResults[i] = rescueRun();
    }

    auto crimeCol = [&](auto field) {
        std::vector<long long> col(RUNS);
        for (int i = 0; i < RUNS; i++) col[i] = crimeResults[i].*field;
        return median(col);
    };

    auto rescueCol = [&](auto field) {
        std::vector<long long> col(RUNS);
        for (int i = 0; i < RUNS; i++) col[i] = rescueResults[i].*field;
        return median(col);
    };

    printf("{\n");
    printf("  \"language\": \"cpp\",\n");
#ifdef __linux__
    printf("  \"os\": \"Linux\",\n");
#elif __APPLE__
    printf("  \"os\": \"macOS\",\n");
#else
    printf("  \"os\": \"unknown\",\n");
#endif
    printf("  \"runtime\": \"g++ C++17\",\n");
    printf("  \"runs\": %d,\n", RUNS);
    printf("  \"crime\": {\n");
    printf("    \"call\": %lld,\n", crimeCol(&CrimeResult::call));
    printf("    \"field\": %lld,\n", crimeCol(&CrimeResult::field));
    printf("    \"calc\": %lld,\n", crimeCol(&CrimeResult::calc));
    printf("    \"single\": %lld,\n", crimeCol(&CrimeResult::single_));
    printf("    \"cache\": %lld,\n", crimeCol(&CrimeResult::cache));
    printf("    \"time\": %lld\n", crimeCol(&CrimeResult::time_));
    printf("  },\n");
    printf("  \"rescue\": {\n");
    printf("    \"call\": %lld,\n", rescueCol(&RescueResult::call));
    printf("    \"arg\": %lld,\n", rescueCol(&RescueResult::arg));
    printf("    \"calc\": %lld,\n", rescueCol(&RescueResult::calc));
    printf("    \"ret\": %lld\n", rescueCol(&RescueResult::ret));
    printf("  }\n");
    printf("}\n");
    return 0;
}

File: harness/kotlin/harness.kt
/**
 * Trace harness: captures median nanosecond costs for the six crime
 * operations and four rescue operations used in the honestcode.software
 * demo engine.
 *
 * Usage:
 *   kotlinc harness.kt -include-runtime -d harness.jar && java -jar harness.jar
 */
import kotlin.system.measureNanoTime

const val WARMUP = 200
const val RUNS = 1000

// ═══════════════════════════════════════════════
// CRIME SCENE: mutable class with singletons
// ═══════════════════════════════════════════════

class CouponRegistry private constructor() {
    companion object {
        @Volatile private var instance: CouponRegistry? = null
        fun getInstance(): CouponRegistry =
            instance ?: synchronized(this) {
                instance ?: CouponRegistry().also { instance = it }
            }
        fun reset() { instance = null }
    }
    fun lookup(code: String): Double = if (code == "SAVE10") 0.10 else 0.0
}

class TaxService private constructor() {
    companion object {
        @Volatile private var instance: TaxService? = null
        fun getInstance(): TaxService =
            instance ?: synchronized(this) {
                instance ?: TaxService().also { instance = it }
            }
        fun reset() { instance = null }
    }
    fun calculate(region: String, taxable: Double): Double = taxable * 0.08
}

data class Item(val name: String, val price: Double)

class Order {
    val items = mutableListOf<Item>()
    var total = 0.0
    var discount = 0.0
    var tax = 0.0
    var couponCode: String? = null
    var updatedAt: Long? = null
}

fun crimeRun(): LongArray {
    val order = Order()
    val items = listOf(Item("Widget", 29.99), Item("Gadget", 39.99), Item("Doohickey", 19.99))

    val callNs = measureNanoTime { order.items.addAll(items) }
    val fieldNs = measureNanoTime { order.couponCode = "SAVE10"; order.discount = 0.0 }
    val calcNs = measureNanoTime { order.total = order.items.sumOf { it.price } }

    CouponRegistry.reset(); TaxService.reset()
    var registry: CouponRegistry? = null
    var taxService: TaxService? = null
    val singleNs = measureNanoTime {
        registry = CouponRegistry.getInstance()
        taxService = TaxService.getInstance()
    } / 2

    val cacheNs = measureNanoTime {
        val rate = registry!!.lookup(order.couponCode!!)
        order.discount = order.total * rate
    }
    val timeNs = measureNanoTime { order.updatedAt = System.nanoTime() }

    order.tax = taxService!!.calculate("NY", order.total - order.discount)

    return longArrayOf(callNs, fieldNs, calcNs, singleNs, cacheNs, timeNs)
}

// ═══════════════════════════════════════════════
// RESCUE: pure functions, flat data
// ═══════════════════════════════════════════════

data class OrderResult(val total: Double, val tax: Double, val subtotal: Double)
data class FinalResult(val total: Double, val tax: Double, val subtotal: Double,
                       val couponCode: String, val discount: Double, val grandTotal: Double)

fun calculateOrder(items: List<Item>, region: String, taxRates: Map<String, Double>): OrderResult {
    val total = items.sumOf { it.price }
    val tax = total * (taxRates[region] ?: 0.0)
    return OrderResult(total, tax, total + tax)
}

fun applyCoupon(order: OrderResult, code: String, coupons: Map<String, Double>): FinalResult {
    val rate = coupons[code] ?: 0.0
    val discount = order.total * rate
    return FinalResult(order.total, order.tax, order.subtotal, code, discount, order.subtotal - discount)
}

fun rescueRun(): LongArray {
    val items = listOf(Item("Widget", 29.99), Item("Gadget", 39.99), Item("Doohickey", 19.99))
    val taxRates = mapOf("NY" to 0.08, "CA" to 0.0725)
    val coupons = mapOf("SAVE10" to 0.10)

    val callNs = measureNanoTime { }
    val argNs = measureNanoTime { @Suppress("UNUSED_VARIABLE") val r = "NY" }
    var result: OrderResult? = null
    val calcNs = measureNanoTime { result = calculateOrder(items, "NY", taxRates) }
    var final_: FinalResult? = null
    val retNs = measureNanoTime { final_ = applyCoupon(result!!, "SAVE10", coupons) }
    check(final_!!.grandTotal > 0)

    return longArrayOf(callNs, argNs, calcNs, retNs)
}

// ═══════════════════════════════════════════════
// HARNESS
// ═══════════════════════════════════════════════

fun median(values: LongArray): Long {
    values.sort()
    return values[values.size / 2]
}

fun main() {
    repeat(WARMUP) { crimeRun(); rescueRun() }

    val crimeResults = Array(RUNS) { crimeRun() }
    val rescueResults = Array(RUNS) { rescueRun() }

    val crimeOps = listOf("call", "field", "calc", "single", "cache", "time")
    val rescueOps = listOf("call", "arg", "calc", "ret")

    val crime = crimeOps.mapIndexed { i, op ->
        op to median(LongArray(RUNS) { crimeResults[it][i] })
    }.toMap()

    val rescue = rescueOps.mapIndexed { i, op ->
        op to median(LongArray(RUNS) { rescueResults[it][i] })
    }.toMap()

    val os = "${System.getProperty("os.name")} ${System.getProperty("os.version")}"
    val runtime = "${System.getProperty("java.vm.name")} ${System.getProperty("java.version")}"

    println(buildString {
        appendLine("{")
        appendLine("""  "language": "kotlin",""")
        appendLine("""  "os": "$os",""")
        appendLine("""  "runtime": "$runtime",""")
        appendLine("""  "runs": $RUNS,""")
        appendLine("""  "crime": {""")
        crime.entries.forEachIndexed { i, (k, v) ->
            appendLine("""    "$k": $v${if (i < crime.size - 1) "," else ""}""")
        }
        appendLine("  },")
        appendLine("""  "rescue": {""")
        rescue.entries.forEachIndexed { i, (k, v) ->
            appendLine("""    "$k": $v${if (i < rescue.size - 1) "," else ""}""")
        }
        appendLine("  }")
        appendLine("}")
    }.trimEnd())
}
